### PR TITLE
Some minor optimizations for speeding up nearest neighbor resize.

### PR DESCRIPTION
**Pull request by panmari**
By removing loop over channels. Impact mainly visible for more than 3
channels. Numbers for benchmark with 6 channels:

```
Before:
Benchmark                                        Time(ns) Iterations
--------------------------------------------------------------------
BM_Resize_ResizeNearestNeighbor_cpu_10_499_499   53222200        100  280.7M items/s

After:
Benchmark                                        Time(ns) Iterations
--------------------------------------------------------------------
BM_Resize_ResizeNearestNeighbor_cpu_10_499_499   44504450        100  335.7M items/s
```

### CODE REVIEW

File: tensorflow/core/kernels/resize_nearest_neighbor_op.cc [hunk 1]
```diff
@@ -56,21 +56,27 @@ class ResizeNearestNeighborOp : public OpKernel {
...
+        const T* input_y_ptr = input_b_ptr + in_y * in_row_size;
         for (int x = 0; x < st.out_width; ++x) {
           const int64 in_x =
               std::min(static_cast<int64>(floorf(x * st.width_scale)),
                        (st.in_width - 1));
-          for (int c = 0; c < st.channels; ++c) {
-            output_data(b, y, x, c) = input_data(b, in_y, in_x, c);
-          }
+          const T* input_yx_ptr = input_y_ptr + in_x * st.channels;
+          output_data_ptr = std::copy_n(input_yx_ptr, st.channels, output_data_ptr);
```
Comment (jhseu) [inline tensorflow/core/kernels/resize_nearest_neighbor_op.cc]: Just call memcpy instead?
```c++
memcpy(&output_data(b, y, x, 0), &input_data(b, in_y, in_x, 0), sizeof(T)*st.channels);
```

See here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/slice_op.cc#L154

**Comment by jhseu** [comment]
Can you also add the BUILD rule for this benchmark? Doesn't seem to be there.

**Comment by jhseu** [comment]
Alternatively, you can use std::copy_n there instead of memcpy too.

**Comment by jhseu** [comment]
Two other things: can you also benchmark on GPU, and can you do best of 5 runs?

**Comment by panmari** [comment]
My change doesn't touch any code that runs on GPU. Nevertheless, I ran the full benchmark 5 times on master and with this patched, see https://gist.github.com/panmari/4b495af0c383c26c1c50435d055c05a5. New version is consistently about 10% faster.

**Comment by panmari** [comment]
Rebased on master.

**Comment by jhseu** [comment]
There's a much simpler version that gets the performance benefit:
```c++
    for (int b = 0; b < st.batch_size; ++b) {
      for (int y = 0; y < st.out_height; ++y) {
        const int64 in_y =
            std::min(static_cast<int64>(floorf(y * st.height_scale)),
                     (st.in_height - 1));
        for (int x = 0; x < st.out_width; ++x) {
          const int64 in_x =
              std::min(static_cast<int64>(floorf(x * st.width_scale)),
                       (st.in_width - 1));
          std::copy_n(&input_data(b, in_y, in_x, 0), st.channels, &output_data(b, y, x, 0));
        }
      }
    }
```

**Review by jhseu — CHANGES_REQUESTED** [review]
See my comment above.

**Comment by panmari** [comment]
Thanks for the hint! The simplified version you proposed gave very similar benchmark results, so I adapted my pull request accordingly.

**Review by jhseu — APPROVED** [review]

**Comment by jhseu** [comment]
Jenkins, test this please

**Pull request merged**